skbeam/core/correlation.py

```diff
@@ -1027,55 +1062,74 @@ def __call__(self, img1, img2=None, normalization=None):
+                             "expected {}".format(self.shape)
+                             )
+        # reshape for 1D case
+        if self.ndim == 1:
+            img1 = img1.reshape((1, self.shape[0]))
+
         if img2 is None:
             self_correlation = True
-            img2 = img1
         else:
             self_correlation = False
+            if img2.shape != self.shape:
+                raise ValueError("Second image not expected shape. " +
+                                 "Got {}".format(img2.shape) +
+                                 " expected {}".format(self.shape))
+            if self.ndim == 1:
+                img2 = img2.reshape((1, self.shape[0]))
 
         ccorrs = list()
         rngiter = tqdm(range(self.nids))
 
+        idpos = self.idpos
         for i in rngiter:
-            self.tmpimgs[i] *= 0
-            self.tmpimgs[i].ravel()[
-                                    self.subpxlsts[i]
-                                    ] = img1.ravel()[self.pxlsts[i]]
+            index_start, index_stop = idpos[i], idpos[i+1]
+            ppiis = self.ppii[index_start:index_stop]
+            ppjjs = self.ppjj[index_start:index_stop]
+            pis = self.pi[index_start:index_stop]
+            pjs = self.pj[index_start:index_stop]
+            tmpimg = np.zeros(self.shapes[i, :])
+            tmpimg[ppiis, ppjjs] = img1[pis, pjs]
+
             if not self_correlation:
-                self.tmpimgs2[i] *= 0
-                self.tmpimgs2[i].ravel()[
-                                        self.subpxlsts[i]
-                                        ] = img2.ravel()[self.pxlsts[i]]
+                tmpimg2 = np.zeros_like(tmpimg)
+                tmpimg2[ppiis, ppjjs] = img2[pis, pjs]
+                tmpimg2[ppiis, ppjjs] = img2[pis, pjs]
 
-            # multiply by maskcorrs > 0 to ignore invalid regions
             if self_correlation:
-                ccorr = _cross_corr(self.tmpimgs[i])*(self.maskcorrs[i] > 0)
+                ccorr = _cross_corr(tmpimg)
             else:
-                ccorr = _cross_corr(self.tmpimgs[i], self.tmpimgs2[i]) * \
-                        (self.maskcorrs[i] > 0)
+                ccorr = _cross_corr(tmpimg, tmpimg2)
 
-            # now handle the normalizations
+            # Note, in this code, non-overlapping regions will now get np.nan
+            # also, for sym averaging, if Icorr*Icorr2==0, then we also get
+            # np.nan
             if 'symavg' in normalization:
                 # do symmetric averaging
-                Icorr = _cross_corr(self.tmpimgs[i] *
-                                    self.submasks[i], self.submasks[i])
+                Icorr = _cross_corr(tmpimg * self.submasks[i],
+                                    self.submasks[i])
                 if self_correlation:
-                    Icorr2 = _cross_corr(self.submasks[i], self.tmpimgs[i] *
+                    Icorr2 = _cross_corr(self.submasks[i], tmpimg *
```

Comment:
not exactly the same; One is reverse of the other, this is very important for symmetric averaging.
However, instead of running convolution again, we could do:
```python
Icorr2 = np.copy(Icorr[::-1, ::-1])
```
(I'm not sure if copy is necessary, I'll have to check)
I have to think about it to be certain and test just in case... (But I'm sure this is equivalent)
maybea cleaner? thanks ill check later :)